### PR TITLE
Stonemary 10-16 week

### PR DESCRIPTION
**Pull request by JesseZhuang**
modify PR to merge into dev branch instead of master

### CODE REVIEW

File: data_types/collections/defaultdict.py [hunk 1]
```diff
@@ -2,7 +2,7 @@
 
 # usage of defaultdict and & in sets
 
-from collections import defaultdict
+from data_types.collections import defaultdict
```
Comment (JesseZhuang) [inline data_types/collections/defaultdict.py]:
``` python
Traceback (most recent call last):
  File "data_types/collections/defaultdict.py", line 5, in <module>
    from data_types.collections import defaultdict
ImportError: No module named 'data_types'
```

File: data_types/collections/namedtuple.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    # keywords: there are 33 of them, in `keyword.kwlist`
+    'class',
+    'for',
+    'return',
+    'global',
+    'pass',
+    'raise'
+):
+    try:
+        BadTypename = namedtuple(bad_typename, 'type shape')
```
Comment (JesseZhuang) [inline data_types/collections/namedtuple.py:50]:
``` python
Traceback (most recent call last):
  File "data_types/collections/namedtuple.py", line 50, in <module>
    BadTypename = namedtuple(bad_typename, 'type shape')
  File "/home/jesse/virtualenvs/iclp3/lib/python3.4/collections/__init__.py", line 345, in namedtuple
    'identifiers: %r' % name)
ValueError: Type names and field names must be valid identifiers: 'Animal!'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "data_types/collections/namedtuple.py", line 52, in <module>
    message = 'Got error for bad_typename={}: {}'.format(bad_typename, error.message)
AttributeError: 'ValueError' object has no attribute 'message'
```

**Review by JesseZhuang — COMMENTED** [review]
Also namedtuple.py has multiple pep8 errors.

**Comment by stonemary** [comment]
probably because we have inconsistent PYTHONPATH. Will take a look tonight.

**Comment by stonemary** [comment]
it was not due to PYTHONPATH, it was due to my IDE python version (in python3, use `str(exception)` rather than `exception.message` and my careless import refactoring. It should be fixed now. Sorry for committing broken code. Thanks for checking.

**Comment by JesseZhuang** [comment]
pep8 line too long errors?

**Comment by stonemary** [comment]
Sorry I usually ignore that one because we all have wide screens now. I can
fix that.

**Pull request merged**